benchmark/run/branched_filter.ex

```diff
@@ -7,17 +7,77 @@ defmodule Benchmark.Run.BranchedFilter do
   def_input_pad :input, accepted_format: _any, availability: :on_request
   def_output_pad :output, accepted_format: _any, availability: :on_request
 
-  def_options number_of_reductions: [spec: integer()]
+  def_options number_of_reductions: [spec: integer()],
+              generator: [spec: (integer() -> integer())],
+              dispatcher: [spec: (integer(), integer() -> [integer()])]
 
   @impl true
   def handle_init(_ctx, opts) do
     workload_simulation = Reductions.prepare_desired_function(opts.number_of_reductions)
-    {[], %{workload_simulation: workload_simulation}}
+
+    {[],
+     %{
+       buffers: [],
+       workload_simulation: workload_simulation,
+       generator: opts.generator,
+       dispatcher: opts.dispatcher
+     }}
   end
 
   @impl true
-  def handle_buffer(_pad, buffer, _ctx, state) do
+  def handle_buffer(_pad, buffer, ctx, state) do
+    state = %{state | buffers: state.buffers ++ [buffer]}
     state.workload_simulation.()
-    {[forward: buffer], state}
+    how_many_buffers_to_output = state.generator.(length(state.buffers))
+
+    if how_many_buffers_to_output > 0 do
+      output_pads = Map.keys(ctx.pads) |> Enum.filter(&match?({Membrane.Pad, :output, _}, &1))
+
+      how_many_buffers_per_pad =
+        state.dispatcher.(length(output_pads), how_many_buffers_to_output)
+
+      {buffers_to_output, rest_buffers} = Enum.split(state.buffers, how_many_buffers_to_output)
+      actions = prepare_actions(buffers_to_output, output_pads, how_many_buffers_per_pad)
+      state = %{state | buffers: rest_buffers}
+      {actions, state}
+    else
+      {[], state}
+    end
+  end
+
+  defp prepare_actions(buffers, output_pads, how_many_buffers_per_pad) do
+    {actions, rest_of_buffers} =
+      Enum.zip(output_pads, how_many_buffers_per_pad)
+      |> Enum.map_reduce(buffers, fn {pad, how_many_buffers_per_pad}, buffers_left ->
+        {buffers_for_this_pad, rest_buffers} = Enum.split(buffers_left, how_many_buffers_per_pad)
+        action = {:buffer, {pad, buffers_for_this_pad}}
+        {action, rest_buffers}
+      end)
+
+    if rest_of_buffers != [] do
+      raise("The dispatcher function is working improperly!")
+    end
+
+    actions
```

Comment:
Second argument of `Enum.zip/2` should be enumerable, but `how_many_buffers_per_pad` sounds like it is a number

```suggestion
Enum.chunk_every(buffers, how_many_buffers_per_pad)
|> Enum.zip(output_pads)
|> Enum.map(fn {buffers_chunk, pad} -> {:buffer, {pad, buffers_chunk}} end)
```

Reply:
Well, `how_many_buffers_per_pad` is not an integer - it's a list of integers which sums to the number of buffers to be dispatched. [1, 4, 3] is a valid `how_many_buffers_per_pad` list for 3 pads and 8 buffers, and such a list means, that one buffer should be sent through the first pad, 4 buffers should be sent through the second pad and 3 buffers should be sent through the third pad.